master/pkg/schemas/defaults.go

```diff
@@ -0,0 +1,233 @@
+package schemas
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// RuntimeDefaultable means there are runtime values for filling in an object, like choosing a
+// random seed based on the wall clock.
+type RuntimeDefaultable interface {
+	// RuntimeDefaults must apply the runtime-defined default values.
+	RuntimeDefaults()
+}
+
+// Defaultable means a struct can have its defaults filled in automatically.
+type Defaultable interface {
+	// DefaultSource must return a parsed json-schema object in which to find defaults.
+	DefaultSource() interface{}
+}
+
+// FillDefaults will recurse through structs, maps, and slices, setting default values for any
+// struct fields whose struct implements the Defaultable interface.  This lets us read default
+// values out of json-schema automatically.
+//
+// There are some forms of defaults which must be filled at runtimes, such as giving a default
+// description to experiments with no description.  This can be accomplished by implementing
+// the RuntimeDefaultable interface for that object.  See ExperimentConfig for an example.
+//
+// There are some objects which get their defaults from other objects' defaults.  This an
+// unforutnate detail of our union types which have common members that appear on the root union
+// object.  That's hard to reason about, and we should avoid doing that in new config objects.  But
+// those objects implement DefaultSource() to customize that behavior.
+//
+// Example usage:
+//
+//    config, err := expconf.ParseAnyExperimentConfigYAML(bytes)
+//
+//    // Use the cluster checkpoint storage if the user did not specify one.
+//    schemas.Merge(&config.CheckpointStorage, cluster_default_checkpoint_storage)
+//
+//    // Define any remaining undefined values.
+//    schemas.FillDefaults(&config)
+//
+func FillDefaults(x interface{}) {
+	obj := reflect.ValueOf(x)
+	if obj.Kind() != reflect.Ptr {
+		panic("FillDefaults must be called on a pointer")
```

Comment:
I think all of these panics can panics can only occur with genuine bugs in the code.  The unmarshaling of defaultBytes, for instance, should absolutely always fail identically; the defaultBytes come from generated code.

(I did go ahead and add a test for unmarshaling defaults which can be applied to all of the generated structs to reduce the chances that any user is the first to find such a bug bug)

For this line which you have flagged, if you had generics and you could replace FillDefaults with:
```go
func WithDefaults[T any](obj T) T {}
```
then this panic would be solved with a compile-time check.  But if this panic ever fires, it will always fire because there's a bug in the code.

Reply:
Ok, sgtm. I agree that they seem like they would indicate genuine bugs, but the difference between errors and panics is that with errors consumers of this library will immediately see the library can error whereas with panics they may be surprised. But panics are fine since I think this will only every be consumed in places where a panic would be handled OK (like an API endpoint or an actor).